Update slide code to be exception safe using let_with_async_scope and try/catch

jesswong opened the pull request.

jesswong committed:
Remove counting_scope::joined(), counting_scope::join_started(), counting_scope::use_count()

ispeters committed:
Update R2 description

Tweak the straw-poll markup to render better as a table and add
backticks around the names of the removed methods.

ispeters committed:
Update paper number to D3149R2

ispeters committed:
Update the description of counting_scope::nest()

Update `counting_scope::nest()` to explain when the scope's count of
outstanding senders is decremented.

jesswong committed:
Add poll notes

jesswong committed:
Fix poll table formatting

jesswong committed:
Add a fourth state to counting_scope

jesswong committed:
Update formatting

ispeters committed:
Finalize P3149R2

 - Tweak the formatting of the revision history for R2
 - Ensure that `counting_scope::nest()` still provides the Strong
   Exception Guarantee in the face of the new "unused" state
 - Make some minor editorial tweaks to clarify wording now that the
   state machine is more complicated

jesswong committed:
Update paper to R3 and make async_scope examples exception safe

jesswong committed:
Merge branch 'p3149r3' into d3149r3

jesswong committed:
Remove merge conflict cruft

ispeters commented on asyncscope.md:
```diff
@@ -275,8 +278,12 @@ int main() {
-    }
+        try {
+	    ex::spawn(std::move(snd), scope); // NEW!
+    	} catch (Exception) {
```
This should probably use `...`
```suggestion
    	} catch (...) {
```

ispeters commented on asyncscope.md:
```diff
@@ -323,7 +330,11 @@ int main() {
-  ex::spawn(std::move(snd), scope);
+  try {
+      ex::spawn(std::move(snd), scope);
+  } catch (Exception) {
```
Same
```suggestion
  } catch (...) {
```

ispeters commented on asyncscope.md:
```diff
@@ -359,39 +370,35 @@ Examples of use
     ex::scheduler auto sch = ctx.scheduler();
 
-    ex::sender auto val = ex::on(sch, ex::just() | ex::then([&result, sch, &scope]() {
+    ex::sender auto val = ex::on(sch, ex::just() | ex::let_with_async_scope([sch](auto scope) {
```
I've been imagining `let_with_async_scope` as taking a callable and no predecessor, in which case this needs to remove `just()`.
```suggestion
    ex::sender auto val = ex::on(sch, ex::let_with_async_scope([sch](auto scope) {
```

ispeters commented on asyncscope.md:
```diff
@@ -359,39 +370,35 @@ Examples of use
             std::cout << "Hello world! Have an int with value: " << val << "\n";
         });
-
+        
```
Nit about the extra whitespace.
```suggestion

```

ispeters commented on asyncscope.md:
```diff
@@ -359,39 +370,35 @@ Examples of use
 }
 
-// The counting scope ensured that all work is safely joined, so result contains 13
+// 'let_with_async_scope' ensured that all work is completed, so result contains 13
```
Should we update the wording to address the possibility that the `spawn` on line 391 threw? Maybe mention that the `sync_wait` will throw whatever is thrown by the callable passed to `let_with_async_scope`?

ispeters commented on asyncscope.md:
```diff
@@ -428,7 +435,11 @@ int main() {
-    my_window window{ctx.get_scheduler(), scope};
+    try {
+        my_window window{ctx.get_scheduler(), scope};
+    } catch (Exception) {
```
```suggestion
    } catch (...) {
```

ispeters commented on asyncscope.md:
```diff
@@ -438,9 +449,7 @@ int main() {
-In this example we use the `counting_scope` within lexical scope to construct an algorithm that performs parallel work.
-This uses the `let_value_with` [@letvwthunifex] algorithm implemented in [@libunifex] which simplifies in-place
-construction of a non-moveable object in the `let_value_with` algorithm's _`operation-state`_ object. Here `foo`
+In this example we use `let_value_with_async_scope` to construct an algorithm that performs parallel work. Here `foo`
```
We need to pick between `let_value_with_async_scope` and `let_with_async_scope` (I prefer the latter because it's shorter; I wonder what @anthonywilliams thinks).
```suggestion
In this example we use `let_with_async_scope` to construct an algorithm that performs parallel work. Here `foo`
```

ispeters commented on asyncscope.md:
```diff
@@ -450,17 +459,13 @@ namespace ex = std::execution;
 ex::sender auto foo(ex::scheduler auto sch) {
-    return unifex::let_value_with([]() noexcept { return ex::counting_scope{}; },
-                   [sch](ex::counting_scope& scope) {
+    return unifex::let_value_with_async_scope([sch](ex::counting_scope scope) {
```
Two things: naming consistency, and the callable's argument is no longer a whole scope, but just a token, so maybe we update to `auto` for now?
```suggestion
    return unifex::let_with_async_scope([sch](auto scope) {
```

ispeters commented on asyncscope.md:
```diff
@@ -494,7 +499,11 @@ task<size_t> listener(int port, io_context& ctx, static_thread_pool& pool) {
-        ex::spawn(scope, std::move(snd));
+        try {
+	    ex::spawn(scope, std::move(snd));
+	} catch (Exception) {
```
```suggestion
	} catch (...) {
```

ispeters reviewed (changes requested):
Looks good, but I have a few thoughts.

ispeters committed:
Add example based on rsys

jesswong committed:
Address comments

jesswong committed:
Merge pull request #8 from ispeters/add_rsys_example

Add example based on rsys

ispeters commented on asyncscope.md:
```diff
@@ -359,39 +370,35 @@ Examples of use
 }
 
-// The counting scope ensured that all work is safely joined, so result contains 13
+// 'let_with_async_scope' ensures that if all work is completed successfully, the result will be 13
```
Nit: I think the "if" is a parenthetical.
```suggestion
// 'let_with_async_scope' ensures that, if all work is completed successfully, the result will be 13
```

ispeters commented on asyncscope.md:
```diff
@@ -494,7 +499,11 @@ task<size_t> listener(int port, io_context& ctx, static_thread_pool& pool) {
                                 return handle_connection(data);
                               });
-        ex::spawn(scope, std::move(snd));
+        try {
+	    ex::spawn(scope, std::move(snd));
+	} catch (...) {
+            // do something with exception
+	}
```
I think this is actually a case where `let_with_async_scope` is better.
```suggestion
      co_await ex::let_with_async_scope([&](auto scope) -> task<void> {
        while (!ctx.is_stopped()) {
          // Accept a new connection
          connection conn = co_await async_accept(ctx, listen_sock);
          count++;
          // Create work to handle the connection in the scope of `work_scope`
          conn_data data{std::move(conn), ctx, pool};
          ex::sender auto snd = ex::just(std::move(data)) |
                                ex::let_value([](auto& data) {
                                  return handle_connection(data);
                                });
          ex::spawn(scope, std::move(snd));
        }
      });
```

jesswong committed:
Update asyncscope.md

Co-authored-by: Ian Petersen <ispeters@meta.com>

jesswong committed:
Update asyncscope.md

Co-authored-by: Ian Petersen <ispeters@meta.com>

jesswong committed:
Fix listener loop in HTTP server example

ispeters commented on asyncscope.md:
```diff
@@ -481,20 +486,21 @@ namespace ex = std::execution;
-                              });
-        ex::spawn(scope, std::move(snd));
+
+      co_await ex::let_with_async_scope([&](auto scope) -> task<void> {
+        while (!ctx.is_stopped()) {
+          // Accept a new connection
+          connection conn = co_await async_accept(ctx, listen_sock);
+          count++;
+          // Create work to handle the connection in the scope of `work_scope`
+          conn_data data{std::move(conn), ctx, pool};
+          ex::sender auto snd = ex::just(std::move(data)) |
+                                ex::let_value([](auto& data) {
+                                  return handle_connection(data);
+                                });
+          ex::spawn(scope, std::move(snd));
+        }
+      });
```
The indentation and bracketing looks off here; I *think* the `co_await` is indented six spaces while the rest of the containing method is indented by four? Also, I think the closing brace on line 504 is extra, right?

jesswong committed:
Fix indentation

ispeters committed:
Split async_scope into async_scope and async_scope_token; add let_with_async_scope (#9)

* Split the concepts, add let_with_async_scope decl

* Small updates to refined list of concepts

* Finish (I think) splitting the concepts

I think this diff completes the process of splitting the concepts; I
need to double-check that references to things like `nest(snd, scope)`
are all updated to `nest(snd, token)`, though.

* More edits

I'm making a pass over the document; I've updated whitespace, tried to
ensure APIs are used correctly, used tokens where appropriate, etc.

* Finish splitting scopes and tokens

* Address @jesswong's feedback

jesswong committed:
Fix exception code to use let_error with counting_scope spawn

jesswong committed:
Counting scope vs let with async (#11)

* Add blurb about counting_scope vs let_with_async_scope

* fix type

* Fix syntax bug

* Fix typos

* Reference P3296R0 for let_with_async_scope

* Fix formatting

* Resolve some PR feedback

Simplify part of the Tony Table comparing "old" code, code with a
`counting_scope`, and code with `let_with_async_scope`.

* Add manual citation for P3296

---------

Co-authored-by: Ian Petersen <ispeters@meta.com>

jesswong committed:
Add examples for recursively spawned work using let_with_async_scope and counting_scope (#10)

* Add example for recursively spawned work using let_with_async_scope and counting_scope

* Update r3 info

* Fix wording

* Address comments

* Address comments and add spawn_future example with error surfacing

* Fix formatting

* Fix formatting

* PR feedback and formatting

---------

Co-authored-by: Ian Petersen <ispeters@meta.com>

ispeters committed:
Add counting_scope::close() and simple_counting_scope (#12)

* Add counting_scope::close() and simple_counting_scope

Incorporate feedback from the folks working on P2300 and make the
following changes:
 - give `counting_scope` a `.close()` method, which makes `.join()` not
   responsible for closing the scope
 - give `counting_scope` an internal stop source and a public
   `request_stop()` method, like `unifex::v1::async_scope`
 - introduce `simple_counting_scope`, which is like `counting_scope` but
   without the stop source or the `.request_stop()` method

This first change only adds a call to `close()` in the rsys-based
example and adds the new type and methods to section 5.1 Definitions;
more change will come in this branch.

* Describe simple_counting_scope

This diff updates section 5.8 to describe `simple_counting_scope`
instead of `counting_scope`. We'll refer to section 5.8 when describing
`counting_scope` in section 5.9.

* Describe counting_scope in terms of simple_counting_scope

This diff gives a brief description of `counting_scope` in terms of
`simple_counting_scope` and `inplace_stop_source`. I'm breaking for
dinner here; I plan to add method-by-method descriptions.

* Finish section 5

This diff describes the methods of `counting_scope`, adds a little more
context to each state in the state machine diagram, and adds the missing
description for `simple_counting_scope::close()`.

* PR feedback

ispeters commented on asyncscope.md:
```diff
@@ -270,13 +280,17 @@ int main() {
-        // and `ctx`)
-        ex::spawn(std::move(snd), scope); // NEW!
-    }
+     // start `snd` as before, but associate the spawned work with `scope` so that it can
+     // be awaited before destroying the resources referenced by the work (i.e. `my_pool`
+     // and `ctx`)
+     ex::spawn(std::move(snd), scope.get_token()); // NEW!
```
This somehow snuck outside the `for` loop.
